### PR TITLE
Remove compacting of TxIn. Add Keyed instances for hashes

### PR DESCRIPTION
**Pull request by lehins**
This PR requires: https://github.com/input-output-hk/cardano-base/pull/243

### CODE REVIEW

**Comment by lehins** [comment]
@redxaxder Here is the reason why we need to remove compacted `TxIn`, we need an instance of `Keyed` for `TxId`

@TimSheard if you are ok with this PR feel free to "Squash and Merge| it, it is opened into your `ts-compact` branch.

**Comment by lehins** [comment]
After some experiments we found out that using HashMap that was implemented in #2520 is qyuite a bit more memory efficient then other approaches that involved manual unpacking of `TxIn`. Here is a comparison:

```
  Case                                  Max        MaxOS         Live       Allocated     GCs
  IntMap (KeyMap TxId ())       232,127,304  503,316,480  232,127,304  42,208,026,560  40,753
  KeyMap TxId (IntMap TxId ())  305,991,112  627,048,448  305,991,112  46,528,677,952  44,916
  IntMap (Map TxId ())          324,725,624  701,497,344  324,725,624  34,313,290,760  33,026
  Map TxIn ()                   383,746,832  896,532,480  383,746,832  33,154,708,560  31,937
```

In order for us to be able and use the most efficient from the top of the list, this PR is needed.

File: libs/cardano-ledger-core/src/Cardano/Ledger/Hashes.hs [hunk 1]
```diff
@@ -75,13 +76,11 @@ newtype ScriptHash crypto
...
-  CC.Crypto crypto =>
-  ToCBOR (ScriptHash crypto)
+deriving newtype instance CC.Crypto crypto => Keyed (ScriptHash crypto)
 
-deriving newtype instance
-  CC.Crypto crypto =>
-  FromCBOR (ScriptHash crypto)
+deriving newtype instance CC.Crypto crypto => ToCBOR (ScriptHash crypto)
+
+deriving newtype instance CC.Crypto crypto => FromCBOR (ScriptHash crypto)
```
Comment (nc6) [inline libs/cardano-ledger-core/src/Cardano/Ledger/Hashes.hs:83]: Where's the underlying `Keyed` instance coming from?

Comment (nc6) [reply libs/cardano-ledger-core/src/Cardano/Ledger/Hashes.hs:83]: Ah, added in this PR :+1:

File: libs/cardano-ledger-core/src/Cardano/Ledger/TxIn.hs [hunk 2]
```diff
@@ -72,101 +71,59 @@ newtype TxId crypto = TxId {_unTxId :: SafeHash crypto EraIndependentTxBody}
...
-  TxInCompact32 ::
-    HS.SizeHash (CC.HASH crypto) ~ 32 =>
-    {-# UNPACK #-} !Word64 -> -- Hash part 1/4
-    {-# UNPACK #-} !Word64 -> -- Hash part 2/4
-    {-# UNPACK #-} !Word64 -> -- Hash part 3/4
-    {-# UNPACK #-} !Word64 -> -- Hash part 4/4
-    {-# UNPACK #-} !Word64 -> -- Index
-    TxIn crypto
-  TxInCompactOther :: !(TxId crypto) -> {-# UNPACK #-} !Word64 -> TxIn crypto
+data TxIn crypto = TxInCompact !(TxId crypto) {-# UNPACK #-} !Int
```
Comment (nc6) [inline libs/cardano-ledger-core/src/Cardano/Ledger/TxIn.hs:87]: Why the change to `Int` here?

**Review by nc6 — APPROVED** [review]
Everything looks reasonable. I have two questions, but they're not really review comments:

- The comment about needing a `Keyed` instance doesn't really motivate the non-unpacking.
- Why the change from `Word64` to `Int` for the index? Is this motivated by wanting to use `IntMap`?

File: libs/cardano-ledger-core/src/Cardano/Ledger/TxIn.hs (hunk 2 above)
Comment (lehins) [reply libs/cardano-ledger-core/src/Cardano/Ledger/TxIn.hs:87]: Because we will use it with `IntMap`, which needs an `Int`, rather than a `Word64`. I don't think it would a problem, right? Considering that max value on mainnet is 249 and I don't see tx index going larger than `maxBound :: Int` any time in the foreseeable future :smile:

File: libs/cardano-ledger-core/src/Cardano/Ledger/Hashes.hs (hunk 1 above)
Comment (lehins) [reply libs/cardano-ledger-core/src/Cardano/Ledger/Hashes.hs:83]: Yep :+1:
